hw04-bytecode/src/main/java/IoC.java

```diff
@@ -0,0 +1,54 @@
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.*;
+
+class IoC {
+
+    static Object create(Object instance) {
+        InvocationHandler handler = new LoggerInvocationHandler(instance);
+        return Proxy.newProxyInstance(IoC.class.getClassLoader(), instance.getClass().getInterfaces(), handler);
+    }
+
+
+    static class LoggerInvocationHandler implements InvocationHandler {
+
+        private final Object instance;
+        private final Map<String, Method> methodsMap;
+
+        LoggerInvocationHandler(Object instance) {
+            this.instance = instance;
+            this.methodsMap = new HashMap<>();
+
+            for (Method method : this.instance.getClass().getMethods()) {
+                methodsMap.put(LoggerInvocationHandler.toShortString(method), method);
+            }
+        }
+
+        private static String toShortString(Method method) {
+            StringBuilder sb = new StringBuilder("method ");
+            sb.append(method.getName());
+            sb.append('(');
+            StringJoiner sj = new StringJoiner(",");
+            for (Class<?> parameterType : method.getParameterTypes()) {
+                sj.add(parameterType.getTypeName());
+            }
+            sb.append(sj);
+            sb.append(')');
+            return sb.toString();
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+            Method clazzMethod = this.methodsMap.get(LoggerInvocationHandler.toShortString(method));
+
+            if (clazzMethod.getAnnotation(Log.class) != null) {
+                System.out.println("executed method: " + method.getName() + ", params: " + Arrays.asList(args));
```

Comment:
Если добавить `@Log` к методу без параметров:
```java
public class UserAction implements IUserAction {
   ...
    @Log
    public void myActionEmpty(){
        System.out.println("empty action");
    }
}
```
то будет NPE:
```
Exception in thread "main" java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:222)
	at java.base/java.util.Arrays$ArrayList.<init>(Arrays.java:4344)
	at java.base/java.util.Arrays.asList(Arrays.java:4331)
	at IoC$LoggerInvocationHandler.invoke(IoC.java:47)
	at com.sun.proxy.$Proxy0.myActionEmpty(Unknown Source)
	at Demo.main(Demo.java:7)
```

Преобразовывать в список лишнее, в Arrays есть метод toString()
```suggestion
                System.out.println("executed method: " + method.getName() + ", params: " + Arrays.toString(args));
```
Но ситуацию отсутствия параметров лучше отдельно обработать.